lib/collection/src/shards/transfer/shard_transfer.rs

```diff
@@ -56,6 +62,86 @@ impl ShardTransfer {
     }
 }
 
+/// Methods for transferring a shard from one node to another.
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Default, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ShardTransferMethod {
+    /// Stream all shard records in batches until the whole shard is transferred.
+    #[default]
+    StreamRecords,
+    /// Snapshot the shard, transfer and restore it on the receiver.
+    Snapshot,
+}
+
+#[allow(clippy::too_many_arguments)]
+pub async fn transfer_shard(
+    transfer_config: ShardTransfer,
+    shard_holder: Arc<LockedShardHolder>,
+    collection_id: CollectionId,
+    collection_name: &str,
+    peer_id: PeerId,
+    channel_service: ChannelService,
+    snapshots_path: &Path,
+    temp_dir: &Path,
+    stopped: Arc<AtomicBool>,
+) -> CollectionResult<()> {
+    let shard_id = transfer_config.shard_id;
+    let transfer_method = transfer_config.method.unwrap_or_default();
+
+    // Initiate shard on a remote peer
+    let remote_shard = RemoteShard::new(
+        shard_id,
+        collection_id.clone(),
+        peer_id,
+        channel_service.clone(),
+    );
+
+    remote_shard.initiate_transfer().await?;
+
+    {
+        let shard_holder_guard = shard_holder.read().await;
+        let transferring_shard = shard_holder_guard.get_shard(&shard_id);
+        if let Some(replica_set) = transferring_shard {
+            match transfer_method {
+                ShardTransferMethod::StreamRecords => {
+                    replica_set.proxify_local(remote_shard.clone()).await?;
+                }
+                ShardTransferMethod::Snapshot => {
+                    replica_set
+                        .queue_proxify_local(remote_shard.clone())
+                        .await?;
+                }
+            }
+        } else {
+            return Err(CollectionError::service_error(format!(
+                "Shard {shard_id} cannot be proxied because it does not exist"
+            )));
+        }
```

Comment:
Why not move `proxify_local`/`queue_proxify_local` calls inside `transfer_batches`/`transfer_snapshot` functions?

Seems weird to:
- do two consecutive matches on `transfer_method`
- with two `shards_holder.read()` calls
  - first to `*proxify_local`
  - and second one is inside `transfer_*` functions
- and then also make an explicit check that `*proxify_local` was called inside `transfer_*`

---

Also:

```suggestion
        let Some(replica_set) = transferring_shard else {
            return Err(CollectionError::service_error(format!(
                "Shard {shard_id} cannot be proxied because it does not exist"
            )));
        };

        match transfer_method {
            ShardTransferMethod::StreamRecords => {
                replica_set.proxify_local(remote_shard.clone()).await?;
            }
            ShardTransferMethod::Snapshot => {
                replica_set
                    .queue_proxify_local(remote_shard.clone())
                    .await?;
            }
        }
```